### ISSUE TITLE
EVP_pkey_print_private printing public key intentional?

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#22233]
The documentation (and the function's name) appear to state that only the private key part should be printed out when calling `EVP_pkey_print_private`:

> The functions EVP_PKEY_print_public(), EVP_PKEY_print_private() and EVP_PKEY_print_params() print out the public, private or parameter components of key pkey respectively. 

This code https://github.com/openssl/openssl/blob/8ed76c62b5d3214e807e684c06efd69c6471c800/crypto/evp/p_lib.c#L1201-L1207

seems to initiate the print-out of private _and_ public key (key pair) though.

Is this a bug in the implementation or a (possible) misinterpretation of the documentation?

When re-reading [the documentation](https://github.com/openssl/openssl/blob/master/doc/man3/EVP_PKEY_print_private.pod) this statement seems to indicate the function is truly _meant_ to print public _and_ private key part:

> For example passing a public key to EVP_PKEY_print_private() will only print the public components.

Would it thus not be more sensible to state that `EVP_pkey_print_private` truly prints the keypair (private _and_ public key) -- or adapt the code to truly request printing only the private key part, i.e., replace "EVP_PKEY_KEYPAIR" with "OSSL_KEYMGMT_SELECT_PRIVATE_KEY")?

**Comment by levitte** [comment]
Yeah, I would characterize this as a bug.  Also, it would be reasonable to have this definition in include/oepnssl/evp.h for this sort of case:

``` C
# define EVP_PKEY_PRIVATE_KEY                                               \
    ( EVP_PKEY_KEY_PARAMETERS | OSSL_KEYMGMT_SELECT_PRIVATE_KEY )
```

**Comment by levitte** [comment]
The confusing bit with all this is that it's unclear what "private key components" are, unless you know the spec for the key type well.  For example, most descriptions would say that a minimum public RSA key is composed of *n* and *e*, while a minimum private RSA key is composed of *n* and *d*.  And yet, the [ASN.1 RSAPrivateKey](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.1.2) definition includes *n*, *e* and *d*.

When the manual talks about "private key components", it really talks about the components of the private key structure.  For RSA keys, that's effectively the full keypair, because that's what the RSAPrivateKey structure looks like.  This has led to some confusion...

**Comment by levitte** [comment]
To be noted is that our 3.0 key2text encoders behave differently than our 1.1.1 ameth backends.  The 3.0 encoders are more true to what one would intuitively think of private vs public components regardless of the structure that holds them, while the 1.1.1 backends are more true to the specification structure.

So the EVP_PKEY_KEYPAIR selector is there to ensure that the printouts that everyone is used to is maintained.  With that in mind, this choice would probably not be regarded as a bug.

However, this is problematic for provided encoder implementations that really do want to print the private key components only.  Do note that you can actually do that, i.e. if both `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` and `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` are set, you can choose to ignore the latter...  but I understand that this feels weird, and for good reasons, as that's compensating for something that's incorrectly specified to begin with.

**Comment by levitte** [comment]
Fix in #22237

**Comment by levitte** [comment]
@baentsch, I'd suggest seeing if the fix (now pushed to master [currently 3.2 alpha], 3.1 and 3.2) works for you.  Sorry that I didn't nudge you earlier.

**Comment by baentsch** [comment]
> @baentsch, I'd suggest seeing if the fix (now pushed to master [currently 3.2 alpha], 3.1 and 3.2) works for you. Sorry that I didn't nudge you earlier.

It does. The problem had been eliminated already for `oqsprovider` by putting in a [guard](https://github.com/open-quantum-safe/oqs-provider/blob/fab30c7f4bb40898aab2a967cdd6b0f990091778/oqsprov/oqs_encode_key2any.c#L1185) that's sensible anyway, though. So, thanks for the fix (in the name of all future provider developers) @levitte!
